File: ios/WatchGuru/Sources/Data/Offline/MutationQueue.swift
import Foundation

/// The changes waiting to reach the server, in the order they were made.
///
/// Persisted, because the case this exists for is a user who marks three
/// episodes on a train and then closes the app. Losing those is worse than
/// never having accepted them — which is also why this one file is written to
/// Documents rather than Caches: the system must not reclaim it.
final class MutationQueue: @unchecked Sendable {

    private static let fileName = "pending-mutations.json"

    private let store: OfflineStore
    private let lock = NSLock()

    init(store: OfflineStore) {
        self.store = store
    }

    /// Appends a mutation and returns it with its assigned id.
    ///
    /// Collapsing happens here rather than at replay time so the queue never
    /// grows unboundedly offline: toggling one episode forty times leaves one
    /// entry, not forty. Safe only because these operations are absolute rather
    /// than relative — "unmark episode 5" does not depend on what preceded it.
    @discardableResult
    func enqueue(_ build: (Int64) -> PendingMutation) -> PendingMutation {
        lock.withLock {
            let current = load()
            let mutation = build((current.map(\.id).max() ?? 0) + 1)
            save(current.filter { $0.target != mutation.target } + [mutation])
            return mutation
        }
    }

    func pending() -> [PendingMutation] {
        lock.withLock { load().sorted { $0.id < $1.id } }
    }

    func remove(id: Int64) {
        lock.withLock { save(load().filter { $0.id != id }) }
    }

    func clear() {
        lock.withLock { store.delete(Self.fileName) }
    }

    private func load() -> [PendingMutation] {
        guard let data = store.read(Self.fileName) else { return [] }
        guard let decoded = try? JSONDecoder().decode([PendingMutation].self, from: data) else {
            // Written by a version that knew different mutation cases.
            // Dropping it loses work, which is bad; retrying for ever against a
            // decoder that cannot read it is worse, and wedges every later sync.
            store.delete(Self.fileName)
            return []
        }
        return decoded
    }

    private func save(_ mutations: [PendingMutation]) {
        guard !mutations.isEmpty else {
            store.delete(Self.fileName)
            return
        }
        guard let data = try? JSONEncoder().encode(mutations) else { return }
        store.write(Self.fileName, data)
    }
}

File: android/app/src/main/kotlin/dev/dhuelin/watchguru/data/offline/MutationQueue.kt
package dev.dhuelin.watchguru.data.offline

import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/**
 * The changes waiting to reach the server, in the order they were made.
 *
 * Persisted, because the case this exists for is a user who marks three
 * episodes on a train and then closes the app. Losing those is worse than
 * never having accepted them.
 *
 * Synchronised rather than lock-free: enqueue happens on whatever thread a
 * screen is on, drain happens on a background one, and the cost of a lock on an
 * operation this rare is not worth reasoning about.
 */
class MutationQueue(
    private val files: FileStore,
    private val json: Json,
) {

    private val lock = Any()

    /**
     * Appends a mutation and returns it with its assigned id.
     *
     * Collapsing happens here rather than at drain time so the queue never
     * grows unboundedly while offline: toggling one episode watched and
     * unwatched forty times leaves one entry, not forty.
     */
    fun enqueue(build: (Long) -> PendingMutation): PendingMutation = synchronized(lock) {
        val current = load()
        val mutation = build((current.maxOfOrNull { it.id } ?: 0L) + 1)

        // A later mutation on the same target supersedes the earlier one. This
        // is only safe because these operations are absolute, not relative:
        // "unmark episode 5" does not depend on what came before it. A
        // relative operation -- "increment", say -- could never be collapsed
        // this way.
        val kept = current.filterNot { it.target == mutation.target }
        save(kept + mutation)
        mutation
    }

    fun pending(): List<PendingMutation> = synchronized(lock) { load().sortedBy { it.id } }

    fun remove(id: Long) = synchronized(lock) {
        save(load().filterNot { it.id == id })
    }

    fun clear() = synchronized(lock) { files.delete(FILE) }

    private fun load(): List<PendingMutation> {
        val raw = files.read(FILE) ?: return emptyList()
        return runCatching {
            json.decodeFromString(ListSerializer(PendingMutation.serializer()), raw)
        }.getOrElse {
            // Unreadable means a queue written by a version of the app that
            // knew different mutation types. Dropping it loses work, which is
            // bad; retrying forever against a parser that cannot read it is
            // worse, and would wedge every future sync.
            files.delete(FILE)
            emptyList()
        }
    }

    private fun save(mutations: List<PendingMutation>) {
        if (mutations.isEmpty()) {
            files.delete(FILE)
        } else {
            files.write(FILE, json.encodeToString(ListSerializer(PendingMutation.serializer()), mutations))
        }
    }

    private companion object {
        const val FILE = "pending-mutations.json"
    }
}
